### PR TITLE
Module 3

### PR DESCRIPTION
**Pull request by StephenSo**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,105 +5,128 @@
...
+
+$ cp .env.template .env # (first time only)
+
+``` 
+
+## Install test dependencies
+
+```
+
+pip3 install pytest selenium python-dotenv flask flask_wtf
```
Comment (JackMead) [inline README.md:42]: We shouldn't need to ask the user to manually install any of these dependencies - any of these required should be tracked in our pyproject.toml. Now that you've added those dependencies, users should be able to install these automatically with `poetry install` and then run `poetry run pytest` to run the tests in an environment that has those dependencies available

File: tests/Integration_test.py [hunk 2]
```diff
@@ -69,30 +144,59 @@ def mock_edit_item(http_method,url,params,headers):
...
+            "name": "Module 01",
+            "idList":"603b8faf893153146df47463"}
+        #json_payload = '{"id":"603b8fd8acd0617a9083538c","checkItemStates":[],"closed":false,"dateLastActivity":"2021-04-05T14:41:54.021Z","desc":"Git","descData":{"emoji":{}},"dueReminder":null,"idBoard":"603b8faf893153146df47462","idList":"603b8faf893153146df47463","idMembersVoted":[],"idShort":3,"idAttachmentCover":null,"idLabels":[],"manualCoverAttachment":false,"name":"Module 01","pos":65535,"shortLink":"JxpPj9Pj","isTemplate":false,"cardRole":null,"dueComplete":false,"due":null,"email":null,"labels":[],"shortUrl":"https://trello.com/c/JxpPj9Pj","start":null,"url":"https://trello.com/c/JxpPj9Pj/3-module-01","idMembers":[],"idChecklists":[],"badges":{"attachmentsByType":{"trello":{"board":0,"card":0}},"location":false,"votes":0,"viewingMemberVoted":false,"subscribed":false,"fogbugz":"","checkItems":0,"checkItemsChecked":0,"checkItemsEarliestDue":null,"comments":0,"attachments":0,"description":true,"due":null,"dueComplete":false,"start":null},"subscribed":false,"cover":{"idAttachment":null,"color":null,"idUploadedBackground":null,"size":"normal","brightness":"light","idPlugin":null}}'
```
Comment (JackMead) [inline tests/Integration_test.py:152]: Let's delete any commented out lines

File: tests/ViewModel_test.py [hunk 3]
```diff
@@ -1,42 +1,59 @@
-import pytest, datetime
-
+import pytest
```
Comment (JackMead) [inline tests/ViewModel_test.py:1]: These tests look much more tightly defined now, nice

File: tests/tests_e2e/test_sel.py [hunk 4]
```diff
@@ -1,52 +1,139 @@
...
+    # start the app in its own thread.
+    thread = Thread(target=lambda:application.run(use_reloader=False))
+    thread.daemon = True
+    thread.start()
+    yield application
+    #Get test List
+    test_list_id = find_list(board_id,'To Do') # all three lists exist already!
+    
+@pytest.fixture(scope="module")
+def app_addItem():
```
Comment (JackMead) [inline tests/tests_e2e/test_sel.py:39]: Here you're creating the item through your test setup - instead we should be doing that within the test itself and we can then combine the setup steps back into a single fixture

File: tests/tests_e2e/test_sel.py [hunk 5]
```diff
@@ -1,52 +1,139 @@
...
 
-def test_task_journey(driver, app_with_temp_board):
+def test_step1_task_createBoard(driver, app_with_temp_board):
     driver.get('http://localhost:5000/')
     assert driver.title == 'To-Do App'
+    
+def test_step2_task_addItem(driver, app_addItem):
+    driver.get('http://localhost:5000/')
+    driver.refresh()
+    assert driver.find_element_by_partial_link_text(test_item_name)
```
Comment (JackMead) [inline tests/tests_e2e/test_sel.py:131]: An example of how we could use selenium in this test might be:

```
    driver.get('http://localhost:5000/NewItem.html')
    elem = driver.find_element_by_id('newItem')
    elem.send_keys(test_item_name)
    submit = driver.find_element_by_xpath("//input[@type='submit']")
    submit.click()
    assert driver.find_element_by_partial_link_text(test_item_name)
```

I suggest having a play with that; can you get a similar setup to work for one of the others below?

**Review by JackMead — CHANGES_REQUESTED** [review]
Thanks for making those changes - I've left a few thoughts below but I think it's just worth fixing up the end-to-end tests; we should interact with our page rather than just getting our test setup to do that. I've left some other thoughts below that would also be worth taking a look at - feel free to reach out if you have any further questions.

Also worth noting that you can leave the pull request open rather than closing it and opening a new one; when you make any further changes it will update this PR (as long as it's to the same branch) and we can review it again once you've resubmitted it to Aptem

File: tests/tests_e2e/test_sel.py (hunk 4 above)
Comment (StephenSo) [reply tests/tests_e2e/test_sel.py:39]: Took me a moment to get this but I did!

File: tests/tests_e2e/test_sel.py (hunk 5 above)
Comment (StephenSo) [reply tests/tests_e2e/test_sel.py:131]: All done.
Alex was online today to help me finish of the last few bits.  You two really do know your stuff!

File: tests/tests_e2e/test_sel.py (hunk 4 above)
Comment (JackMead) [reply tests/tests_e2e/test_sel.py]: Yeah, it's a subtle change from a code-writing perspective, but it makes a big difference in how useful our test is!

**Review by JackMead — APPROVED** [review]
Thanks for making those changes - your end-to-end test now looks really solid, happy for you to merge this when you're ready 😄

**Pull request merged**